### ISSUE TITLE
run H2GCN alone on graph with heterophily

### ISSUE THREAD

**Issue by ShunliRen** [GemsLab/H2GCN#4]
Sorry to bother you with another question: How can I run H2GCN alone on graph with heterophily? After downloading them, how to load them without signac?

**Comment by ShunliRen** [comment]
Or could you provide an example that loading dataset use signac? eg. to load cora with h=0.1
THX!

**Comment by jiong-zhu** [comment]
With signac, you can achieve this by adding more constraints in the JSON config. For example, to load syn-cora with only homophily ratio h=0.1, you could modify the `graph_filter_dict` object in [this JSON config](https://github.com/GemsLab/H2GCN/blob/master/experiments/h2gcn/configs/syn-cora/h2gcn.json) to follows:
```json
 "graph_filter_dict": {
         "numNode": 1490, 
         "h.$where": "lambda h: abs(h - 0.1) < 1e-5" 
     }
```
More usages of the filters supported can be found [here](https://docs.signac.io/projects/core/en/latest/api.html#signac.Collection.find). After that you can again follow the command as described [here](https://github.com/GemsLab/H2GCN#replicate-experiments-with-signac):
```bash
python run_hgcn_experiments.py -c configs/syn-cora/h2gcn.json [-i] run [-p PARALLEL_NUM] 
```
I hope these information would be helpful, and please let me know if you have any additional questions!

**Comment by ShunliRen** [comment]
Thanks!
I have another question: Could I use a GAT to replace the GCN in H2GCN?   Will the performance of the model be improved by replacing GCN with GAT ?If yes, where should I rewrite the code?

**Comment by ShunliRen** [comment]
And what does 'adjhops' in the code mean?

**Comment by ShunliRen** [comment]
And what does the ''layerTag'' mean in the layer? How does it work?

**Comment by jiong-zhu** [comment]
> I have another question: Could I use a GAT to replace the GCN in H2GCN? Will the performance of the model be improved by replacing GCN with GAT ?If yes, where should I rewrite the code?

If you want to try out your own architecture, you can add your own Python modules to the folder `/h2gcn/models`. One easier way to do so is to folk our H2GCN module and make your own modifications. Once you have the module ready, you can invoke it using signac by replacing the `H2GCN` in `model_args` object in [this JSON config](https://github.com/GemsLab/H2GCN/blob/master/experiments/h2gcn/configs/syn-cora/h2gcn.json) to the name of your own module (i.e., the python file name under `/h2gcn/models`). We don't feel adding attention mechanisms help as much, if at all, as the designs we identify in our paper under heterophily settings. 




> And what does the ''layerTag'' mean in the layer? How does it work?

Basically, `layerTag` names the output of a specific layer, which can serve as input to deeper layers of the architecture. For example, we use `layerTag` along with the [`ConcatLayer`](https://github.com/GemsLab/H2GCN/blob/6e42eec34d854c1adf9e01c2c26985d03262d9bc/h2gcn/models/_layers.py#L83)to implement jumping-knowledge connections by concatenating tagged outputs from previous layers with the current outputs).

**Comment by ShunliRen** [comment]
thanks for your answers!
what does ''adjhops'' look like? Where is it defined?

**Comment by ShunliRen** [comment]
It seems that  I can't run the code on GPU without CUDA11.0, is that right?

**Comment by ShunliRen** [comment]
> thanks for your answers!
> what does ''adjhops'' look like? Where is it defined?

for example, the shape of adjhop matrix for citeseer is [3327,2,3327], what does it mean?

**Comment by jiong-zhu** [comment]
> > thanks for your answers!
> > what does ''adjhops'' look like? Where is it defined?
> 
> for example, the shape of adjhop matrix for citeseer is [3327,2,3327], what does it mean?

The `adjhops` variable is a tensor of adjacency matrix **when considering different hop of neighbors**. For example, under the default argument with `--adj_nhood 1 2`, `adjhops[:, 0, :]` corresponds to the 1-hop adjacency matrix, while `adjhops[:, 1, :]` corresponds to the 2-hop adjacency matrix (representing 2-hop neighborhood $$\bar{N}_2(v)$$ in our paper, or $$\bar{A}_2$$ in Algorithm 1 in Appendix D). You can also refer to the source code for calculating this at [Line 551 to 558 of this file](https://github.com/GemsLab/H2GCN/blob/6e42eec34d854c1adf9e01c2c26985d03262d9bc/h2gcn/datasets/_dataset.py#L551).

> It seems that I can't run the code on GPU without CUDA11.0, is that right?

I think as long as you have successfully installed `TensorFlow >= 2.0`, the exact version of CUDA is neglectable. However, you will need to install some compatible version of CUDA if you intend to run TensorFlow >= 2.0 with GPU acceleration.
